### PR TITLE
support for warning(s)  initial commit

### PR DESCRIPTION
**Pull request by Sainyam2003**
#248

### CODE REVIEW

**Review by objorke — CHANGES_REQUESTED** [review]
The submodule must be removed

File: Source/Examples/PropertyGrid/PropertyGridDemo/Examples/NotifyDataErrorInfoSeverityExample.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            get
+            {
+                return this.firstName;
+            }
+
+            set
+            {
+                this.firstName = value;
+                this.Validate("FirstName", !string.IsNullOrEmpty(this.firstName), "First Name should be specified.",Severity.Error);
+                this.Validate("FirstName", !string.IsNullOrEmpty(this.firstName), "First Name should not be numeric.", Severity.Warning);
```
Comment (objorke) [inline Source/Examples/PropertyGrid/PropertyGridDemo/Examples/NotifyDataErrorInfoSeverityExample.cs:45]: use IsNumeric?

File: Source/Examples/PropertyGrid/PropertyGridDemo/CustomControlFactory.cs [hunk 2]
```diff
@@ -31,5 +33,71 @@ public override FrameworkElement CreateControl(PropertyItem pi, PropertyControlF
...
+            }
+            else
+            {
+                errorConverter = new NotifyDataErrorInfoConverter(notifyDataErrorInfoInstance, pi.PropertyName);
+                propertyPath = nameof(tab.HasErrors);
+                source = tab;
+                notifyDataErrorInfoInstance.ErrorsChanged += (s, e) =>
+                {
+                    //tab.UpdateHasErrors(notifyDataErrorInfoInstance);
+                    tab.UpdateTabForValidationResults(notifyDataErrorInfoInstance);
```
Comment (objorke) [inline Source/Examples/PropertyGrid/PropertyGridDemo/CustomControlFactory.cs:64]: can implement directly here (or in a private method)
```
tab.HasErrors = ...
tab.HasWarnings = ...
```

File: Source/PropertyTools.Wpf/Validation/ValidationResultEx.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ValidationResultEx.cs" company="PropertyTools">
+//   Copyright (c) 2014 PropertyTools contributors
+// </copyright>
+// <summary>
+//   An extension for ValidationResult.
+// </summary>
+// --------------------------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
```
Comment (objorke) [inline Source/PropertyTools.Wpf/Validation/ValidationResultEx.cs:10]: the coding style is to put "using" inside the namespace block

**Review by objorke — CHANGES_REQUESTED** [review]
This looks good!

I have suggested a few changes to make less changes on the library, I think it would be better to put all the custom functionality into the example. It is possible that there may be more severity levels, we should not limit the `PropertyGrid` to errors and warnings only.

I created #248 for this PR.

Remember to update the change log and authors/contributors files.

File: Source/Examples/PropertyGrid/PropertyGridDemo/Examples/NotifyDataErrorInfoSeverityExample.cs (hunk 1 above)
Comment (objorke) [reply Source/Examples/PropertyGrid/PropertyGridDemo/Examples/NotifyDataErrorInfoSeverityExample.cs:45]: I don't think the Validate method supports being called two times

**Review by objorke — CHANGES_REQUESTED** [review]

File: Source/PropertyTools.Wpf/Validation/ValidationResultEx.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+using System.ComponentModel.DataAnnotations;
+
+namespace PropertyTools.Wpf
+{
+    /// <summary>
+    /// Adds severity in ValidationResult
+    /// </summary>
+    /// <seealso cref="System.ComponentModel.DataAnnotations.ValidationResult" />
+    public class ValidationResultEx : ValidationResult
```
[the remark this answers is not in the document]
Comment (Sainyam2003) [reply Source/PropertyTools.Wpf/Validation/ValidationResultEx.cs]: done.

File: Source/PropertyTools.Wpf/PropertyGrid/Tab.cs [hunk 5]
```diff
@@ -109,15 +133,37 @@ public void UpdateHasErrors(IDataErrorInfo dei)
...
             // validate all properties in this tab
-            this.HasErrors = this.Groups.Any(g => g.Properties.Any(p => ndei.HasErrors));
+            this.HasErrors = this.Groups.Any(g => g.Properties.Any(p => ndei.HasErrors));            
+        }
+
+        /// <summary>
+        /// Updates the tab for validation results, to include Errors and Warngins both
+        /// </summary>
+        /// <param name="errorInfo">The error information.</param>
+        public void UpdateTabForValidationResults(INotifyDataErrorInfo errorInfo)
```
[the remark this answers is not in the document]
Comment (Sainyam2003) [reply Source/PropertyTools.Wpf/PropertyGrid/Tab.cs]: done.

File: Source/PropertyTools.Wpf/PropertyGrid/PropertyGrid.cs [hunk 6]
```diff
@@ -1622,7 +1624,7 @@ private void ConfigurePropertyControl(PropertyItem pi, FrameworkElement property
         /// <summary>
         /// Creates the error control.
         /// </summary>
-        private ContentControl CreateErrorControl(PropertyItem pi, object instance, Tab tab)
+        public virtual ContentControl CreateErrorControl(PropertyItem pi, object instance, Tab tab)
```
[the remark this answers is not in the document]
Comment (Sainyam2003) [reply Source/PropertyTools.Wpf/PropertyGrid/PropertyGrid.cs:1627]: it is because of inheriting from interface

**Comment by Sainyam2003** [comment]
I would still need your help to change Control's border color in case of warning(s), I can not find the place in code.

**Review by objorke — COMMENTED** [review]
The `PropertyGrid` control looks ok now!
I just cannot get the example to work properly, the error messages are not showing as expected and the color of the tab header is not changing.

File: Source/PropertyTools.Wpf/PropertyGrid/PropertyGrid.cs [hunk 7]
```diff
@@ -1370,7 +1371,8 @@ private void AddPropertyPanel(Panel panel, PropertyItem pi, object instance, Tab
                         propertyControl.Style = this.ValidationErrorStyle;
                     }
 
-                    errorControl = CreateErrorControl(pi, instance, tab);
+                    PropertyControlFactoryOptions options = new PropertyControlFactoryOptions { ValidationErrorTemplate = this.ValidationErrorTemplate };
```
Comment (objorke) [inline Source/PropertyTools.Wpf/PropertyGrid/PropertyGrid.cs:1374]: move this code into the ´PropertyGrid.CreateErrorControl´ method?
or remove the ´PropertyGrid.CreateErrorControl´ method

**Review by objorke — CHANGES_REQUESTED** [review]
The example is still not working properly.
Besides from that it looks good to merge!

**Review by objorke — APPROVED** [review]

**Pull request merged**